### PR TITLE
feat: added geopose and geopose covariance

### PR DESCRIPTION
**Pull request by KyleJewiss**

### CODE REVIEW

File: src/septentrio_gnss_driver/communication/message_handler.cpp [hunk 1]
```diff
@@ -2308,6 +2573,8 @@ namespace io {
             if (settings_->publish_atteuler)
                 publish<AttEulerMsg>("atteuler", last_atteuler_);
             assemblePoseWithCovarianceStamped();
+            assembleGeoPoseStamped();
```
Comment (natrad100) [inline src/septentrio_gnss_driver/communication/message_handler.cpp:2552]: PoseWithCovarainceStamped() has:
- pose
  - position[x,y,x]
  - orientation[x,y,z,w]
- covariance[n]

GeoPoseStamped() has
- pose
    - position[lat,lon,alt]
    - orientation[x,y,z,w]
GeoPoseWithCovarianceStamped() has
- pose
    - position[lat,lon,alt]
  - orientation[x,y,z,w]
- covariance[n]

you can see that there is a pattern here. Think about how you might be able to reuse some of the calculations from these functions, noting that the functions get disabled by the config.

**hint**
```cpp
get_covariance()
geometry_msgs::msg::Point get_pose()
geographic_msgs::msg::GeoPoint get_geopose()
float64[36] get_orientation()
```

Acknowledging that we may want to rebase with this repo later, I still think it's worth doing.

Other option would be to do something like the following which would minimise code change, clean up your own code but looks ugly as hell and wouldn't recomend it.
```cpp
struct RotationWithCovariance{
    orientation: geometry_msgs::msg::Quaternion
    covariance: float64[36]
}
std::pair<geometry_msgs::msg::Quaternion, float64[36]> assemblePoseWithCovarianceStamped()
geometry_msgs::msg::GeoPoint assembleGeoPoseStamped(orientation)
assembleGeoPoseWithCovariance(geopoint, orientation, covariance)
```

File: src/septentrio_gnss_driver/communication/message_handler.cpp [hunk 2]
```diff
@@ -216,9 +217,251 @@ namespace io {
...
+                convertCovariance(last_attcoveuler_.cov_headroll);
+            msg.pose.covariance[34] =
+                convertCovariance(last_attcoveuler_.cov_headpitch);
+            msg.pose.covariance[35] =
+                convertAutoCovariance(last_attcoveuler_.cov_headhead);
+        }
+        publish<GeoPoseWithCovarianceStampedMsg>("geopose_covariance_stamped", msg);
     };
 
+    float MessageHandler::getCovariance()
```
Comment (natrad100) [inline src/septentrio_gnss_driver/communication/message_handler.cpp]: `float MessageHandler::getCovarianceErrorLonLat()`

Comment (natrad100) [inline src/septentrio_gnss_driver/communication/message_handler.cpp]: I was also wrong, the error calculation should be the following: `2 * sqrt(cov_lat + cov_lon)`

Functions should be as unit testable as possible

```cpp
float MessageHandler::getHorizontalError(last_poscovgeodetic_ poscovgeodetic){
    return 2 * sqrt( poscovgeodetic.cov_lonlon + poscovgeodetic.cov_latlat )
}
```

File: src/septentrio_gnss_driver/communication/message_handler.cpp [hunk 3]
```diff
@@ -216,9 +217,251 @@ namespace io {
...
+    }
+
+    bool MessageHandler::covarianceOutsideThreshold()
+    {
+        float covariance = getCovariance();
+        float covariance_threshold;
+        node_->get_parameter("covariance_threshold", covariance_threshold);
+
+        if (covariance > covariance_threshold ||
+            covariance < -covariance_threshold || std::isnan(covariance))
```
Comment (natrad100) [inline src/septentrio_gnss_driver/communication/message_handler.cpp]: this if statements is legit

```cpp
if (something_is_true){
    return true;
}
return false;
```

In cpp, what the following is doing can be thought as
```cpp
// Original
if(covariance > covariance_threshold || covariance < -covariance_threshold || std::isnan(covariance))

// Interpreted
if ( bool or bool or bool )

// Example
if( true or false or false )
// Evals as
if (true)

// thus you can simply
return whatever_the_logic_compairson_is;
```

File: src/septentrio_gnss_driver/communication/message_handler.cpp (hunk 2 above)
Comment (KyleJewiss) [reply src/septentrio_gnss_driver/communication/message_handler.cpp]: does this naming supersede your previous suggestion to name the function `float MessageHandler::getCovarianceErrorLonLat()`

File: src/septentrio_gnss_driver/communication/message_handler.cpp [hunk 4]
```diff
@@ -216,9 +217,251 @@ namespace io {
...
     };
 
+    float MessageHandler::getCovariance()
+    {
+        return sqrt(
+            (last_poscovgeodetic_.cov_lonlon * last_poscovgeodetic_.cov_lonlon) +
+            (last_poscovgeodetic_.cov_latlat * last_poscovgeodetic_.cov_latlat));
+    }
+
+    bool MessageHandler::covarianceOutsideThreshold()
```
Comment (natrad100) [inline src/septentrio_gnss_driver/communication/message_handler.cpp]: this probably should be `horizontalErrorOutsideThreshold()` sorry

File: src/septentrio_gnss_driver/communication/message_handler.cpp [hunk 5]
```diff
@@ -360,7 +603,13 @@ namespace io {
         receiver_status.values[4].key = "CPU load in %";
         receiver_status.values[4].value =
             std::to_string(last_receiverstatus_.cpu_load);
-        if ((last_receiverstatus_.rx_error & (1 << 9)))
+
+        if (covarianceOutsideThreshold())
```
Comment (natrad100) [inline src/septentrio_gnss_driver/communication/message_handler.cpp]: This code block is done 3 times. Move into a function, maybe

```cpp
void applyHorizontalErrorStat(DiagnosticStatusMsg* receiver_status){
    if (horizontalErrorOutsideThreshold()){
        receiver_status.level = DiagnosticStatusMsg::ERROR;
        receiver_status.message = "Covariance outside threshold {" +
                                    std::to_string(getHorizontalError()) + "}";
    }
}
```
This allows us to operate directly on the reciever status object and would be called with
```cpp
// ... other things
applyHorizontalErrorStat(reciever_status);
```

File: src/septentrio_gnss_driver/communication/message_handler.cpp (hunk 2 above)
Comment (natrad100) [reply src/septentrio_gnss_driver/communication/message_handler.cpp]: yes sorry, the value is an error for lat and lon so getHorizontalError is a good name me thinks

File: src/septentrio_gnss_driver/communication/message_handler.cpp [hunk 6]
```diff
@@ -163,30 +171,6 @@ namespace io {
...
-            roll = std::isnan(roll) ? 0.0 : roll;
-            pitch = std::isnan(pitch) ? 0.0 : pitch;
-
-            msg.pose.pose.orientation = convertEulerToQuaternionMsg(
-                deg2rad(roll), deg2rad(pitch), deg2rad(yaw));
-            msg.pose.pose.position.x = rad2deg(last_pvtgeodetic_.longitude);
-            msg.pose.pose.position.y = rad2deg(last_pvtgeodetic_.latitude);
-            msg.pose.pose.position.z = last_pvtgeodetic_.height;
             // Filling in the covariance data in row-major order
             msg.pose.covariance[0] = last_poscovgeodetic_.cov_lonlon;
```
Comment (natrad100) [inline src/septentrio_gnss_driver/communication/message_handler.cpp]:
```suggestion
    // Function Definition
    float64[36] MessageHandler::getGeoPoseCovData( void )

    // Use
    geopose.pose.covariance = getGeoPoseCovData();
```

**Review by natrad100 — APPROVED** [review]
lgtm

**Pull request merged**